### PR TITLE
Add support for foreach over list values

### LINKED ISSUE #120: Add support for foreach over list values
Fallow on from #119 . I think we can just check the type of the collection < List and reduce it to while loop just like range. Also need to update semantic analyzer to validate the variable type

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
$subject 
Resolves #120

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

 

## Summary by CodeRabbit

* **New Features**
  * Foreach loops now support iterating over list collections (index-driven).
  * Multi-dimensional array types resolved across dimensions.

* **Bug Fixes**
  * Safer handling of unspecified/nil array dimensions to avoid traversal issues.
  * Elvis operator now reported as unsupported.

* **Tests**
  * Added many corpus examples for loops, lists, equality, nil checks and integer edge cases; some serialization tests skipped in CI.

* **Style**
  * Array size printing now uses bracketed formatting.

### CODE REVIEW

File: bir/codec/corpus_codec_test.go [hunk 1]
```diff
@@ -45,6 +45,7 @@ func getBIRDiff(expectedText, actualText string) string {
 
 // TestBIRSerialization tests BIR serialization and deserialization roundtrip from .bal source files in the corpus.
 func TestBIRSerialization(t *testing.T) {
+	t.Skip("BIR serialization test disabled")
```
Comment (heshanpadmasiri) [inline bir/codec/corpus_codec_test.go:48]: Remove this after #117

File: corpus/integration_test.go [hunk 2]
```diff
@@ -60,6 +60,7 @@ var (
 	// Skip tests that cause unrecoverable Go runtime errors
 	skipTestsMap = makeSkipTestsMap([]string{
 		"subset3/03-list/24-v.bal",
+		"subset3/03-int/neg-p.bal",
```
Comment (heshanpadmasiri) [inline corpus/integration_test.go:63]: Remove after https://github.com/ballerina-platform/ballerina-lang-go/issues/110

File: semantics/semantic_analyzer.go [hunk 3]
```diff
@@ -904,11 +904,29 @@ func analyzeWhile[A analyzer](a A, whileStmt *ast.BLangWhile) {
...
+		collectionType := collection.GetDeterminedType()
+		var expectedValueType semtypes.SemType
+		switch {
+		case semtypes.IsSubtypeSimple(collectionType, semtypes.LIST):
+			memberTypes := semtypes.ListAllMemberTypesInner(a.tyCtx(), collectionType)
+			var result semtypes.SemType = &semtypes.ANY
+			for _, each := range memberTypes.SemTypes {
+				result = semtypes.Intersect(result, each)
+			}
+			expectedValueType = result
```
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply semantics/semantic_analyzer.go:923]: Fixed

File: desugar/statement.go [hunk 4]
```diff
@@ -306,10 +306,138 @@ func visitForEach(cx *Context, stmt *ast.BLangForeach) desugaredNode[model.State
...
+		BLangAccessExpressionBase: ast.BLangAccessExpressionBase{
+			Expr: collection,
+		},
+		IndexExpr: idxVarRef,
+	}
+	elementAccess.SetDeterminedType(loopVarDef.Var.GetDeterminedType())
+
+	// Step 6: patch loop var def initial expression
+	loopVarDef.Var.SetInitialExpression(elementAccess)
+
```
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply desugar/statement.go:397]: Fixed

[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply desugar/statement.go:397]: Fixed in [b2b5b69](https://github.com/ballerina-platform/ballerina-lang-go/pull/124/commits/b2b5b691d6b51d750a343ab2a5aaa72c7873fe72)

File: projects/module_context.go [hunk 5]
```diff
@@ -264,6 +264,8 @@ func compileInternal(moduleCtx *moduleContext) {
 
 	// Run CFG analyses (reachability and explicit return) after semantic analysis.
 	semantics.AnalyzeCFG(cx, pkgNode, cfg)
+
+	moduleCtx.bLangPkg = desugar.DesugarPackage(moduleCtx.compilerCtx, moduleCtx.bLangPkg, importedSymbols)
```
Comment (warunalakshitha) [inline projects/module_context.go:269]: Should we add a comment here?

Comment (heshanpadmasiri) [reply projects/module_context.go:269]: Fixed

File: desugar/statement.go [hunk 6]
```diff
@@ -306,10 +306,157 @@ func visitForEach(cx *Context, stmt *ast.BLangForeach) desugaredNode[model.State
...
+	whileStmt.SetDeterminedType(&semtypes.NEVER)
+
+	return desugaredNode[model.StatementNode]{
+		initStmts:       initStmts,
+		replacementNode: whileStmt,
+	}
+}
+
+func createLengthInvocation(cx *Context, collection ast.BLangExpression) *ast.BLangInvocation {
+	const pkgName = "lang.array"
```
Comment (warunalakshitha) [inline desugar/statement.go]: should be a global constant?

Comment (heshanpadmasiri) [reply desugar/statement.go]: fixed
